kernel/paging.c

```diff
@@ -0,0 +1,312 @@
+#include "declarations.h"
+#include "functions.h"
+
+// This file contains all the functions which would be
+// help in paging
+
+/*
+    Virtual Address
+    -----------------------------------------------------------
+    |  unused: 25 | L2 (root): 9 | L1: 9 | L0: 9 | offset: 12 |
+    -----------------------------------------------------------
+    L2: Root page directory
+    L1: Second level page directory
+    L0: Final level page directory
+
+    Virtual Page Number
+    ---------------------------------------
+    |  unused: 37 | L2: 9 | L1: 9 | L0: 9 |
+    ---------------------------------------
+    L2, L1 and L0 together determine the Virtual Page Number
+
+    Physical Address
+    -------------------------------------------------------
+    |  unused: 8 | Physical Page Number : 44 | offset: 12 |
+    -------------------------------------------------------
+
+    Page Table Entry
+    -------------------------------------------------------
+    |  unused: 10  | Physcial Page Number: 44 | flags: 10 |
+    -------------------------------------------------------
+*/
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Waddress-of-packed-member"
+
+// Get offset into a page, it is valid for both physical and virtual addresses
+// Get least 12 bits
+uint64 getOffset(uint64 addr) {
+    return addr & 0x000000000000ffff;
+}
```

Comment:
it's 12 or 16 bits. In comments, it's 12 bit but you are extracting 16 bits.

Reply:
Oh it is 12, that is a very good catch !!
I'll correct it thanks !!